app/shared/utils/EmptinessChecker.scala

```diff
@@ -108,50 +108,54 @@ object EmptinessChecker {
     instance(list => Structure.Arr(list.map(aInstance.structureOf)))
+  given [A](using aInstance: EmptinessChecker[A]): EmptinessChecker[Seq[A]] =
+    instance(seq => Structure.Arr(seq.map(aInstance.structureOf)))
 
-  implicit def listInstance[A](implicit aInstance: EmptinessChecker[A]): EmptinessChecker[List[A]] =
-    instance(list => Structure.Arr(list.map(aInstance.structureOf)))
+    // Lazy prevents infinite recursion in generic derivation
+  final class Lazy[+A](val value: () => A) extends AnyVal
+  object Lazy {
+    implicit def make[A](implicit a: => A): Lazy[A] = new Lazy(() => a)
+  }
 
-  implicit val hnilInstance: ObjEmptinessChecker[HNil] = instanceObj(_ => Structure.Obj(Nil))
-
-  implicit def hlistInstance[K <: Symbol, H, T <: HList](implicit
-      witness: Witness.Aux[K],
-      hInstance: Lazy[EmptinessChecker[H]],
-      tInstance: ObjEmptinessChecker[T]
-  ): ObjEmptinessChecker[FieldType[K, H] :: T] =
-    instanceObj { case h :: t =>
-      val hField  = witness.value.name -> hInstance.value.structureOf(h)
-      val tFields = tInstance.structureOf(t).fields
-      Structure.Obj(hField :: tFields)
+  inline given derived[A](using m: Mirror.Of[A]): EmptinessChecker[A] =
+    inline m match {
+      case p: Mirror.ProductOf[A] => productInstance(p)
     }
 
-  implicit def simpleHlistInstance[H, T <: HList](implicit
-      hInstance: Lazy[EmptinessChecker[H]],
-      tInstance: ObjEmptinessChecker[T]
-  ): ObjEmptinessChecker[(String, H) :: T] =
-    instanceObj { case h :: t =>
-      val hField  = h._1 -> hInstance.value.structureOf(h._2)
-      val tFields = tInstance.structureOf(t).fields
-      Structure.Obj(hField :: tFields)
+  private inline def productInstance[A](p: Mirror.ProductOf[A]): EmptinessChecker[A] = {
+    val elemLabels = summonLabels[p.MirroredElemLabels]
+    val elemInstances = summonAllInstances[p.MirroredElemTypes]
+    instance { a =>
+      val elems = a.asInstanceOf[Product].productIterator.toList
+      val fields = elemLabels.zip(elems).zip(elemInstances).map {
+        case ((label, value), checker: Lazy[EmptinessChecker[Any]]) =>
+          label -> checker.value().structureOf(value)
+      }
+      Structure.Obj(fields)
     }
+  }
 
-  implicit def genericInstance[A, R](implicit
-      gen: LabelledGeneric.Aux[A, R],
-      enc: Lazy[EmptinessChecker[R]]
-  ): EmptinessChecker[A] =
-    instance(a => enc.value.structureOf(gen.to(a)))
+  private inline def summonLabels[T <: Tuple]: List[String] =
+    inline erasedValue[T] match {
+      case _: (h *: t) => constValue[h].asInstanceOf[String] :: summonLabels[t]
+      case _: EmptyTuple => Nil
+    }
 
+  private inline def summonAllInstances[T <: Tuple]: List[Lazy[EmptinessChecker[Any]]] =
+    inline erasedValue[T] match {
+      case _: (h *: t) => Lazy.make(using summonInline[EmptinessChecker[h]]).asInstanceOf[Lazy[EmptinessChecker[Any]]] :: summonAllInstances[t]
+      case _: EmptyTuple => Nil
+    }
```

Comment:
Same comments as in the `UnexpectedJsonFieldsValidator`. Please see [here](https://github.com/hmrc/individuals-reliefs-api/commit/8d5ec9fe06ffdad860876b9e2e7d50a544d5c0c4#diff-87784f2edea532d0ed97686273baa945b56eeff15601ece2e2f9088be2bb670b), thanks